common/journald/src/stream.rs

```diff
@@ -0,0 +1,356 @@
+                    Ok(shared_state) => shared_state,
+                    Err(e) => {
+                        // we can't wake up the stream so it will hang indefinitely; need
+                        // to panic here
+                        panic!(
+                            "journald's worker thread unable to access shared state: {:?}",
+                            e
+                        );
+                    }
+                };
+                if let Some(waker) = shared_state.waker.take() {
+                    waker.wake();
+                }
+            };
+
+            loop {
+                let is_alive = match thread_shared_state.lock() {
+                    Ok(shared_state) => shared_state.is_alive,
+                    Err(e) => {
+                        // we can't wake up the stream so it will hang indefinitely; need
+                        // to panic here
+                        panic!(
+                            "journald's worker thread unable to access shared state: {:?}",
+                            e
+                        );
+                    }
+                };
+
+                if !is_alive {
+                    break;
+                } else if let RecordStatus::Line(line) = journal.process_next_record() {
+                    if let Err(e) = sender.send(line) {
+                        warn!(
+                            "journald's worker thread unable to communicate with main thread: {}",
+                            e
+                        );
+                        break;
+                    }
+
+                    call_waker();
+                } else if let Err(e) = journal.reader.wait(Some(Duration::from_millis(100))) {
+                    warn!(
+                        "journald's worker thread unable to poll journald for next record: {}",
+                        e
+                    );
+                    break;
+                }
+            }
+
+            // some sort of error has occurred. Explicitly drop the sender before waking up the
+            // stream to prevent a race condition
+            drop(sender);
+            call_waker();
+        });
+
+        self.thread = Some(thread);
+        self.receiver = Some(receiver);
+    }
+
+    fn drop_thread(&mut self) {
+        self.shared_state.lock().unwrap().is_alive = false;
+        if let Some(thread) = self.thread.take() {
+            if let Err(e) = thread.join() {
+                warn!("unable to join journald's worker thread: {:?}", e)
+            }
+        }
+    }
+}
+
+impl FutureStream for Stream {
+    type Item = Vec<LineBuilder>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let mut self_ = self.as_mut();
+
+        if let Some(ref receiver) = self_.receiver {
+            match receiver.try_recv() {
+                Ok(line) => {
+                    return Poll::Ready(Some(vec![line]));
```

Comment:
I'd love some way to avoid or reuse the vecs we're passing back here, allocating for every return hurts me on some basic level. That's a future problem thought

Reply:
dropped a comment